flax/struct.py

```diff
@@ -162,11 +163,50 @@ def from_state_dict(x, state):
   serialization.register_serialization_state(
       data_clz, to_state_dict, from_state_dict)
 
+  def new_setattr(obj, name, value):
+    if obj.__mutable__:
+      object.__setattr__(obj, name, value)
+    else:
+      raise dataclasses.FrozenInstanceError(f"cannot assign to field '{name}'")
+
+  data_clz.__mutable__ = False # denotes whether the object's currently mutable
+  data_clz.__setattr__ = new_setattr # type: ignore
+
   # add a _flax_dataclass flag to distinguish from regular dataclasses
   data_clz._flax_dataclass = True # type: ignore[attr-defined]
 
   return data_clz # type: ignore
 
+@contextlib.contextmanager
+def create_mutable_copy(obj: _T) -> Generator[_T, None, None]:
+
+  def set_mutable_attribute(obj, value: bool, visited: set):
+    if id(obj) in visited:
+      return
+    visited.add(id(obj))
+
+    if dataclasses.is_dataclass(obj) and hasattr(obj, '__mutable__'):
+      object.__setattr__(obj, '__mutable__', value)
+
+    for child_obj in get_children(obj):
+      set_mutable_attribute(child_obj, value, visited)
+
+  def get_children(obj):
+    registry_entry = _registry.get(type(obj))
```

Comment:
do we have to refer to the jax pytree private registry API, or is there a way to achieve the same with public API calls?